### PR TITLE
chore(core): added Matrix4.fromTranslationQuaternion

### PR DESCRIPTION
**Pull request by jfayot**
Added Matrix4.fromTranslationQuaternion.
And updated Matrix4.getScale, getTranslation, getRotation and getRotationMatrix3 to return same type as input. This allows the following kind of construct:
`new Matrix4().getRotationMatrix3().invert()`

### CODE REVIEW

File: modules/core/src/classes/matrix4.ts [hunk 1]
```diff
@@ -371,7 +389,9 @@ export class Matrix4 extends Matrix {
    * @param result
    * @returns self
    */
-  getTranslation(result: NumericArray = [-0, -0, -0]): NumericArray {
+  getTranslation(): Vector3;
+  getTranslation<T extends NumericArray>(result: T): T;
+  getTranslation(result = new Vector3()): Vector3 {
```
Comment (ibgreen) [inline modules/core/src/classes/matrix4.ts]: Should we change the return type here in the implementation, as it is not always a Vector3?

File: modules/core/test/classes/matrix4.spec.ts [hunk 2]
```diff
@@ -42,6 +42,11 @@ test('Matrix4#fromQuaternion', (t) => {
   t.end();
 });
 
+test('Matrix4#fromTranslationQuaternion', (t) => {
+  tapeEquals(t, new Matrix4().fromTranslationQuaternion([0, 0, 0], [0, 0, 0, 1]), IDENTITY_MATRIX);
```
Comment (ibgreen) [inline modules/core/test/classes/matrix4.spec.ts]: Appreciate the test case. Maybe add one test case with non-zero vectors to test the non-trivial case?

**Review by ibgreen — DISMISSED** [review]
Much appreciated.

File: modules/core/src/classes/matrix4.ts (hunk 1 above)
Comment (jfayot) [reply modules/core/src/classes/matrix4.ts]: Well, I guess we can keep the NumericArray return type in the default case, that still allows to chain calls if a Vector3 is provided as input... I'll update that !

File: modules/core/test/classes/matrix4.spec.ts (hunk 2 above)
Comment (jfayot) [reply modules/core/test/classes/matrix4.spec.ts]: Yes, a litlle bit lazy here !!! Just inspired myself from the test case Matrix4#fromQuaternion. I'll do my best to update that !

File: modules/core/src/classes/matrix4.ts (hunk 1 above)
Comment (ibgreen) [reply modules/core/src/classes/matrix4.ts]: Thanks. My understanding when using overloaded signatures is that the "implementation" signature doesn't become visible to the user. Only the overloaded signatures without implementation become part of the public API. 

If I am correct, the change I am proposing is more a matter of making the code a little clearer to maintain.

File: modules/core/test/classes/matrix4.spec.ts (hunk 2 above)
Comment (ibgreen) [reply modules/core/test/classes/matrix4.spec.ts]: Thanks, admit that our existing tests are not always that good. But that seems like a poor reason for keeping a bad habit in new code....

Comment (jfayot) [reply modules/core/test/classes/matrix4.spec.ts]: Done ! Tell me what you think

File: modules/core/src/classes/matrix4.ts (hunk 1 above)
Comment (jfayot) [reply modules/core/src/classes/matrix4.ts]: Fixed

**Comment by jfayot** [comment]
> Much appreciated.

Hi again ! Sorry, I just realized that this repo has moved to visgl/math.gl.
Do you want me to submit this PR to the new repo and delete this one ?

**Comment by ibgreen** [comment]
> I just realized that this repo has moved to visgl/math.gl

Sorry I din't notice that this was in the uber-web repo. Yes, that would be good. 

(We are having a discussion with Uber to see if we can solve this split but for now we need to work in the vis.gl fork. )
